cpp/arcticdb/storage/s3/mock_s3_client.cpp

```diff
@@ -0,0 +1,166 @@
+        S3Operation operation_to_fail,
+        Aws::S3::S3Errors error_to_fail_with) {
+    return fmt::format("{}#Failure_{}_{}", s3_object_name, operation_to_string(operation_to_fail), (int)error_to_fail_with);
+}
+
+std::optional<Aws::S3::S3Error> has_failure_trigger(const std::string& s3_object_name, S3Operation operation){
+    auto failure_string_for_operation = "#Failure_" + operation_to_string(operation) + "_";
+    auto position = s3_object_name.rfind(failure_string_for_operation);
+    if (position == std::string::npos) return std::nullopt;
+    try {
+        auto failure_code_string = s3_object_name.substr(position + failure_string_for_operation.size());
+        auto failure_code = Aws::S3::S3Errors(std::stoi(failure_code_string));
+        return Aws::S3::S3Error(Aws::Client::AWSError<Aws::S3::S3Errors>(failure_code, "Simulated error", "Simulated error message", true));
+    } catch (std::exception& e) {
+        return std::nullopt;
+    }
+}
+
+const auto not_found_error = Aws::S3::S3Error(Aws::Client::AWSError<Aws::S3::S3Errors>(Aws::S3::S3Errors::RESOURCE_NOT_FOUND, false));
+
+S3Result<std::monostate> MockS3Client::head_object(
+        const std::string& s3_object_name,
+        const std::string &bucket_name) {
+    auto maybe_error = has_failure_trigger(s3_object_name, S3Operation::HEAD);
+    if (maybe_error.has_value()) return {maybe_error.value()};
+
+    if (store.find({bucket_name, s3_object_name}) == store.end()){
+        return {not_found_error};
+    }
+    return {std::monostate()};
+}
+
+
+S3Result<Segment> MockS3Client::get_object(
+        const std::string &s3_object_name,
+        const std::string &bucket_name) {
+    auto maybe_error = has_failure_trigger(s3_object_name, S3Operation::GET);
+    if (maybe_error.has_value()) return {maybe_error.value()};
+
+    auto pos = store.find({bucket_name, s3_object_name});
+    if (pos == store.end()){
+        return {not_found_error};
+    }
+    return {pos->second};
+}
+
+S3Result<std::monostate> MockS3Client::put_object(
+        const std::string &s3_object_name,
+        Segment &&segment,
+        const std::string &bucket_name) {
+    auto maybe_error = has_failure_trigger(s3_object_name, S3Operation::PUT);
+    if (maybe_error.has_value()) return {maybe_error.value()};
+
+    store.insert_or_assign({bucket_name, s3_object_name}, std::move(segment));
+
+    return {std::monostate()};
+}
+
+S3Result<DeleteOutput> MockS3Client::delete_objects(
+        const std::vector<std::string>& s3_object_names,
+        const std::string& bucket_name) {
+    for (auto& s3_object_name : s3_object_names){
+        auto maybe_error = has_failure_trigger(s3_object_name, S3Operation::DELETE);
+        if (maybe_error.has_value()) return {maybe_error.value()};
+    }
+
+    DeleteOutput output;
+    for (auto& s3_object_name : s3_object_names){
+        auto maybe_error = has_failure_trigger(s3_object_name, S3Operation::DELETE_LOCAL);
+        if (maybe_error.has_value())
```

Comment:
Can we avoid unbracketed `if-s` and especially `if-else`s please? They're a bit error prone, and it's annoying to put a breakpoint on a one-line if.